### PR TITLE
Extra meta data remover

### PR DESCRIPTION
**Pull request by mariusmihaic**
Implemented tool that creates transactions to remove all esdt meta data for provided tokens.
For each shard, this tool will:
1. Sort tokensID by nonces. E.g. `token1-rand1-0f`, `token1-rand1-01`, `token1-rand1-02` => `token1-rand1` : `{1,2, 15}`
2. Group nonces in contiguous intervals for each token. E.g. `token1-rand1` : `{1,2, 15}` =>  `token1-rand1` will have 2 intervals:
- `token1-rand1: interval1: {start: 1, end: 2}`
- `token1-rand1:  interval2: {start: 15, end: 15}`
3. Sort tokens by max contiguous nonces intervals. E.g. `interval1` (has 2 nonces) > `interval2` (has 1 nonce)
4. Group token nonces in bulks to be added in tx data. Here, we **prioritze** the longest nonces intervals and **split intervals in such way that we delete the exact number of nonces per each tx**. E.g.: if we want to delete 2 nonces per tx, then we would have:
- first bulk: delete `token1-rand1-01` + `token1-rand1-02`
- second bulk: delete `token1-rand1-0f`
5. Format tx data
6. Sign and generate txs

Input
--

- `Tokens` - these are the tokens that will be deleted. Input should be map<shardID, token>
- `Pems` - For each shard in `Tokens`, a pem must be provided to sign the corresponding shard txs. Each pem should have the name `shard[ID].pem`, where [ID] = 0,1,2... 
- `config.toml`:
1. `ProxyUrl` = needed for network config + account details
2. `TokensToDeletePerTransaction` each generated transaction will delete exactly the same number of token nonces specified by this param
7. `GasLimit` for each tx (should be adjusted based on TokensToDeletePerTransaction`)

Output
--

- For each shard, a json file with signed txs will be generated, e.g.: `txsShard0.json`

TODO (will follow in next PRs)
--
These are not included in this PR, since it is already pretty big in size

- Also save interval state (tokens + intervals), along txs
- Split `txsCreator` functionality in interfaces
- More unit tests
- Maybe add --verbose flag

### CODE REVIEW

File: miscellaneous/go.mod [hunk 1]
```diff
@@ -0,0 +1,49 @@
+module github.com/ElrondNetwork/elrond-tools-go/miscellaneous
```
Comment (iulianpascalau) [inline miscellaneous/go.mod:1]: let's rename this to something more meaningful. `miscellaneous` is way to general. Something like `tokensRemoverTxGenerator` or something like this

File: miscellaneous/metaDataRemover/config.toml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+ProxyUrl = ""
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/config.toml:1]: might add explanatory comments with some examples for testnet/devnet/mainnet

File: miscellaneous/metaDataRemover/config.toml [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+
+TokensToDeletePerTransaction = 0
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/config.toml:3]: why 0?

File: miscellaneous/metaDataRemover/config/config.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+// ContextFlagsMetaDataRemover is the flags config for meta data remover
+type ContextFlagsMetaDataRemover struct {
+	trieToolsCommon.ContextFlagsConfig
+	Outfile string
+	Tokens  string
+	Pems    string
+}
+
+type Config struct {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/config/config.go:13]: missing comment

File: miscellaneous/metaDataRemover/main.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	bytesFromJson, err := ioutil.ReadAll(jsonFile)
+	if err != nil {
+		return nil, err
+	}
+
+	shardTokensMap := make(map[uint32]map[string]struct{})
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/main.go:106]: as in the other PR, this is hard to follow: `map[uint32]map[string]struct{}` might represent anything in any form. Suggestion to change to a proper struct

File: miscellaneous/metaDataRemover/main.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	var cfg config.Config
+	err = toml.Unmarshal(tomlBytes, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cfg, nil
+}
+
+func readPemsData(pemsFile string) (map[uint32]*pkAddress, error) {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/main.go:131]: we need to redesign this. From the experience, we won't have .pems for signing the transactions, it will likely be a ledger behind the mnemonics. Suggestion to somehow format the transaction, prepare the gas price/gas limit, set the value to 0, set the correct address and then, someone else will just sign the transaction on another computer. Maybe @andreibancioiu will guide us how this was done in the erdpy

File: miscellaneous/metaDataRemover/main.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+		return 0, fmt.Errorf("invalid file input name = %s; expected pem file name to be <shardX.pem>, where X = number(e.g. shard0.pem)", file)
+	}
+
+	return uint32(shardID), nil
+}
+
+func getNumTokens(shardTokensMap map[uint32]map[string]struct{}) int {
+	numTokensInShard := 0
+	for _, tokens := range shardTokensMap {
+		for range tokens {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/main.go:179]: please remove this for an instead use:
```GO
numTokensInShard += len(shardTokensMap)
```

File: miscellaneous/metaDataRemover/mocks/proxy.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
+package mocks
+
+import (
+	"context"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+)
+
+type ProxyStub struct {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/mocks/proxy.go:9]: missing comments on exported items

File: miscellaneous/metaDataRemover/mocks/transactionInteractor.go [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package mocks
+
+import (
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+)
+
+type TransactionInteractorStub struct {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/mocks/transactionInteractor.go:7]: missing comments

File: miscellaneous/metaDataRemover/tokenSorting.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+	tokenID  string
+	interval *interval
+}
+
+type tokenData struct {
+	tokenID   string
+	intervals []*interval
+}
+
+func sortTokensIDByNonce(tokens map[string]struct{}) (map[string][]uint64, error) {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/tokenSorting.go:25]: this could have been part of the struct that also held the data. For example, you could have had a data structure `token` like this:
```go
type token struct{
    identifier string
    nonce      uint64
}
```
of course with an `NewToken` constructor function that could have populated both fields

and the map that contains the tokens on a shard could have actually held slice
```go
type tokensOnShard struct{
    tokens  []token
    shardID uint32
}
```

and then, you could have just a sort function call on the slice

File: miscellaneous/metaDataRemover/tokenSorting.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+			numNoncesInBulk = 0
+		}
+
+		index++
+	}
+
+	return bulks
+}
+
+func splitInterval(currInterval *interval, index uint64) (*interval, *interval) {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/tokenSorting.go:161]: these could have been part of the interval struct

File: miscellaneous/metaDataRemover/tokenSorting_test.go [hunk 12]
```diff
@@ -0,0 +1,356 @@
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSortTokensIDByNonce(t *testing.T) {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/tokenSorting_test.go:9]: :+1:

File: miscellaneous/metaDataRemover/txCreator.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/interactors"
+	"github.com/ElrondNetwork/elrond-tools-go/miscellaneous/metaDataRemover/config"
+)
+
+type pkAddress struct {
+	privateKey []byte
+	address    core.AddressHandler
+}
+
+func getPrivateKeyAndAddress(pemFile string) (*pkAddress, error) {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/txCreator.go:26]: why not create a struct for this functionality in order to be able to test it

File: miscellaneous/metaDataRemover/txDataFormater.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
+package main
+
+import (
+	"encoding/hex"
+
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/builders"
+)
+
+const ESDTDeleteMetadataPrefix = "ESDTDeleteMetadata"
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/txDataFormater.go:9]: this can be renamed to function. Also, do we need to be exported?

File: miscellaneous/metaDataRemover/txDataFormater.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+func createShardTxsDataMap(shardTokensMap map[uint32]map[string]struct{}, tokensToDeletePerTx uint64) (map[uint32][][]byte, error) {
```
Comment (iulianpascalau) [inline miscellaneous/metaDataRemover/txDataFormater.go:11]: why not a struct for this functionality?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: miscellaneous/go.mod (hunk 1 above)
Comment (mariusmihaic) [reply miscellaneous/go.mod:1]: Renamed it to `tokensRemover`, since tools inside this file will not only generate txs

File: miscellaneous/metaDataRemover/config/config.go (hunk 4 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/config/config.go:13]: Added it

File: miscellaneous/metaDataRemover/config.toml (hunk 2 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/config.toml:1]: Added explanatory comments

File: miscellaneous/metaDataRemover/config.toml (hunk 3 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/config.toml:3]: Wanted to leave all values to some default values.
However, changed `TokensToDeletePerTransaction` && `GasLimit`

File: miscellaneous/metaDataRemover/main.go (hunk 7 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/main.go:179]: Right, refactored as suggested

File: miscellaneous/metaDataRemover/mocks/proxy.go (hunk 8 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/mocks/proxy.go:9]: Right, added comments for all stubs

File: miscellaneous/metaDataRemover/mocks/transactionInteractor.go (hunk 9 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/mocks/transactionInteractor.go:7]: Added it

File: miscellaneous/metaDataRemover/tokenSorting.go (hunk 10 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/tokenSorting.go:25]: > type token struct{
    identifier string
    nonce      uint64
}

I honestly disagree with this approach.
I don't see a strong benefit of creating a struct that should do on constructor exactly what this func does.
Why over-complicate things when you have an input and need an output? That's the job of a simple func.

> type tokensOnShard struct{

Also disagree here. Every sorting done in `tokenSorting.go` is independent of any shard, they just sort tokens. Also, why would I create a struct wich has `shardID` as a field, when I feel like is more natural to have a map to sort the tokens by shardID ?

File: miscellaneous/metaDataRemover/tokenSorting.go (hunk 11 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/tokenSorting.go:161]: Agree, added this func as part of `interval` struct

File: miscellaneous/metaDataRemover/tokenSorting_test.go (hunk 12 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/tokenSorting_test.go:9]: :pray:

File: miscellaneous/metaDataRemover/main.go (hunk 6 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/main.go:131]: Unfortunately, this would not be feasible, since in case of removing 100 nonces/tx, with a volume of 7_000_000 nonces, we would need to sign 70_000 txs with the ledger

File: miscellaneous/metaDataRemover/txCreator.go (hunk 13 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/txCreator.go:26]: Right, created interface for it and a struct

File: miscellaneous/metaDataRemover/txDataFormater.go (hunk 15 above)
Comment (mariusmihaic) [reply miscellaneous/metaDataRemover/txDataFormater.go:11]: >Split txsCreator functionality in interfaces

As described in PR description, probably a refactor in that area will follow if necessary

Comment (iulianpascalau) [reply miscellaneous/metaDataRemover/txDataFormater.go]: right, missed the description

File: miscellaneous/metaDataRemover/main.go (hunk 6 above)
Comment (iulianpascalau) [reply miscellaneous/metaDataRemover/main.go]: right, will remain as it is. Will figure out how to properly manage this.

**Review by iulianpascalau — APPROVED** [review]

File: tokensRemover/metaDataRemover/txCreator.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+			return fmt.Errorf("no pem data provided for shard = %d", shardID)
+		}
+
+		log.Info("starting to create txs", "shardID", shardID, "num of txs", len(txsData))
+		txsInShard, err := createTxs(pemData, proxy, ti, txsData, cfg.GasLimit)
+		if err != nil {
+			return err
+		}
+
+		file := outFile + "/txsShard" + strconv.Itoa(int(shardID)) + ".json"
```
Comment (bogdan-rosianu) [inline tokensRemover/metaDataRemover/txCreator.go:70]: one useful feature would be to have a configured number of txs per bulk. or even better, a num-nonces-based txs split. For example, we could have txsShard2_bulk0.json, txsShard2_bulk1.json, and so on. each bulk could contain a number of pre-defined txs, or even better, each file will contain a number of nonces (200 for example). This will make the sending of the transactions much easier, since if we would have a file containing all txs, it would be really hard to determine how to send them correctly

File: tokensRemover/metaDataRemover/config/config.go [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+type ContextFlagsMetaDataRemover struct {
+	trieToolsCommon.ContextFlagsConfig
+	Outfile string
+	Tokens  string
+	Pems    string
+}
+
+// Config holds the config for meta data remover tool
+type Config struct {
+	ProxyUrl                     string `toml:"ProxyUrl"`
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply tokensRemover/metaDataRemover/config/config.go:15]: Right, but decided to keep the consistency with `config.toml`, in which it feels more natural to have all fields with uppercase

File: tokensRemover/metaDataRemover/flags.go [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+		Value: "output",
+	}
+	tokens = cli.StringFlag{
+		Name:  "tokens",
+		Usage: "This flag specifies the input file; it expects the input to be a map<shardID, tokens>",
+		Value: "tokens.json",
+	}
+	pems = cli.StringFlag{
+		Name:  "pem",
+		Usage: "This flag specifies pems file which should be used to sign txs. It expects each pem/shardID to be named shard[ID].pem",
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply tokensRemover/metaDataRemover/flags.go:22]: Yes, that's right. Name is indeed unclear, changed it !

File: tokensRemover/metaDataRemover/main.go [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+func startProcess(c *cli.Context) error {
+	flagsConfig := getFlagsConfig(c)
+
+	_, errLogger := trieToolsCommon.AttachFileLogger(log, logFilePrefix, flagsConfig.ContextFlagsConfig)
+	if errLogger != nil {
+		return errLogger
+	}
+
+	log.Info("sanity checks...")
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply tokensRemover/metaDataRemover/main.go:52]: Right, removed it

File: tokensRemover/metaDataRemover/pemProvider.go [hunk 20]
```diff
@@ -0,0 +1,33 @@
+package main
+
+import (
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/interactors"
+)
+
+type pkAddress struct {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply tokensRemover/metaDataRemover/pemProvider.go:8]: No bad naming, pk= privateKey, which is indeed unfortunate here, since pk could also mean public key.
Renamed it to `sk`

File: tokensRemover/metaDataRemover/tokenSorting.go [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+	ret := make(map[string][]uint64)
+	for token := range tokens {
+		splits := strings.Split(token, "-")
+		if len(splits) != 3 {
+			return nil, fmt.Errorf("found invalid format in token = %s; expected [ticker-randSequence-nonce]", token)
+		}
+
+		tokenID := splits[0] + "-" + splits[1] // ticker-randSequence
+		nonceBI := big.NewInt(0)
+		nonceBI.SetString(splits[2], 16)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply tokensRemover/metaDataRemover/tokenSorting.go:49]: Right, added extra check as suggested

File: tokensRemover/metaDataRemover/txCreator.go (hunk 16 above)
Comment (mariusmihaic) [reply tokensRemover/metaDataRemover/txCreator.go:70]: I disagree with this.
The tool that will handle txs sending could format its input however it suits for it. This tool should not hold extra logic to format its output specific to the input of another tool. We simply output all signed txs for each shard, which is is the most complete information to be formatted as pleased from this point forward

> one useful feature would be to have a configured number of txs per bulk.

just to add extra clarification, in the context:  one bulk == one tx, which holds multiple tokens with multiple nonces

> or even better, a num-nonces-based txs split.

No, we do **not** want to choose the number of nonces based on number of txs (this is exactly the reverted logic of how it is implemented now). It does not matter how many txs are created, we only care for consistency in number of nonces to be deleted per tx.

> For example, we could have txsShard2_bulk0.json, txsShard2_bulk1.json, and so on. each bulk could contain a number of pre-defined txs, or even better, each file will contain a number of nonces (200 for example). This will make the sending of the transactions much easier, since if we would have a file containing all txs, it would be really hard to determine how to send them correctly

I'm not sure I understand the benefit of this approach. This adds a lot of overhead. Why not have a tool that, for example:
1. Reads all txs from shardX in memory
2. Simply sends one tx per block (or even multiple, although I doubt it) and keeps an internal counter for each sent tx
3. Store this counter in a file
4. In case anything happens during execution (e.g.: internet connection lost, tool forced stop), we could restart it and start from the last stored index

Comment (bogdan-rosianu) [reply tokensRemover/metaDataRemover/txCreator.go:70]: okay, can remain as it is :+1:

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]
